# Análisis predictivo de datos - Evaluación 1

**Condiciones de entrega de la evaluación:**

- Debe entregar un archivo en formato .ipynb.
- El código debe correr correctamente cuando las celdas se ejecutan de manera secuencial.
- Debe entregar el código ya ejecutado.
- Elimine previamente todo el código que sea basura, solo entregue el código que de respuesta a lo que se le pide que haga.

Con el archivo de datos suministrado, va a crear un modelo de aprendizaje de máquina que prediga la variable **Weight**. Para escoger el modelo a utilizar se debe tener en cuenta que esta variable es **numérica** o **cuantitativa**.

Información sobre este archivo, incluido el diccionario de datos, la puede obtener en [Kaggle](https://www.kaggle.com/datasets/samruddhim/olympics-althlete-events-analysis).

## Carga y comprensión de los datos

Abra el archivo de datos en Excel o bloc de notas, y revise sí hay algo particular con los datos. Este paso es importante para evitar problemas posteriores al cargar los datos.

Cargue en `Pandas` el archivo de datos.

Indique cuántas variables y registros tiene el dataset, y asegúrese que el tipo de dato de cada variable sea el esperado.

Reporte las estadísticas descriptivas de las variables numéricas y categóricas.

Identifique si el dataframe tiene datos duplicados.

Identifique si hay variables con datos nulos.

Reporte los hallazgos en un celda de texto.

#### <span style="color:blue">Cargando el archivo

In [115]:
from pathlib import Path
import pandas as pd

ROOT_DIR = Path().resolve().parent
DATA_DIR = ROOT_DIR / "data/raw"
file_path = 'athlete_events.csv'

df = pd.read_csv(DATA_DIR / file_path)
df.head()

,ID,Name,Sex,Age,Height,Weight,Team,NOC,Games,Year,Season,City,Sport,Event,Medal
0,1,A Dijiang,M,24.0,180.0,80.0,China,CHN,1992 Summer,1992,Summer,Barcelona,Basketball,Basketball Men's Basketball,NaN
1,2,A Lamusi,M,23.0,170.0,60.0,China,CHN,2012 Summer,2012,Summer,London,Judo,Judo Men's Extra-Lightweight,NaN
2,3,Gunnar Nielsen Aaby,M,24.0,NaN,NaN,Denmark,DEN,1920 Summer,1920,Summer,Antwerpen,Football,Football Men's Football,NaN
3,4,Edgar Lindenau Aabye,M,34.0,NaN,NaN,Denmark/Sweden,DEN,1900 Summer,1900,Summer,Paris,Tug-Of-War,Tug-Of-War Men's Tug-Of-War,Gold
4,5,Christine Jacoba Aaftink,F,21.0,185.0,82.0,Netherlands,NED,1988 Winter,1988,Winter,Calgary,Speed Skating,Speed Skating Women's 500 metres,NaN


In [116]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 31.0+ MB


In [117]:
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 31.0+ MB


In [ ]:
df.describe() #Numéricas

,ID,Age,Height,Weight,Year
count,271116.000000,261642.000000,210945.000000,208241.000000,271116.000000
mean,68248.954396,25.556898,175.338970,70.702393,1978.378480
std,39022.286345,6.393561,10.518462,14.348020,29.877632
min,1.000000,10.000000,127.000000,25.000000,1896.000000
25%,34643.000000,21.000000,168.000000,60.000000,1960.000000
50%,68205.000000,24.000000,175.000000,70.000000,1988.000000
75%,102097.250000,28.000000,183.000000,79.000000,2002.000000
max,135571.000000,97.000000,226.000000,214.000000,2016.000000


In [119]:
df.describe(include='object').T #Categoricas

,count,unique,top,freq
Name,271116,134732,Robert Tait McKenzie,58
Sex,271116,2,M,196594
Team,271116,1184,United States,17847
NOC,271116,230,USA,18853
Games,271116,51,2000 Summer,13821
Season,271116,2,Summer,222552
City,271116,42,London,22426
Sport,271116,66,Athletics,38624
Event,271116,765,Football Men's Football,5733
Medal,39783,3,Gold,13372


#### <span style="color:blue">Datos duplicados

In [120]:
print('Número de registros duplicados: ', df.duplicated().sum())

Número de registros duplicados:  1385


#### <span style="color:blue">Datos nulos

In [121]:
df.isnull().sum().sort_values(ascending=False).head(5)

Medal     231333
Weight     62875
Height     60171
Age         9474
Sex            0
dtype: int64

#### <span style="color:blue">El dataset tiene 15 variables y 271116 registros.

#### <span style="color:blue">Las variables Age y Height debe ser tipo int y no float, ya que revisando el archivo no se encuentran valores con decimales. Puede pasar que dice tipo float porque en estás variables hay datos nulos.

#### <span style="color:blue">Los datos numéricos encontrados en el dataset son: ID, Age, Height, Weight y Year.

#### <span style="color:blue"> Los datos numéricos encontrados en el dataset son: Name, Sex, Team, NOC, Games, Season, City, Sport, Event y Medal.

#### <span style="color:blue">Hay 1385 datos duplicados en el dataset.

#### <span style="color:blue">Las variables que tienen datos nulos son Medal-231333, Weight-62875, Height-60171 y Age-9474

Descarte las variables que considere irrelevantes para su modelo (p.e. claves primarias, variables con valores únicos, variables redundantes, entre otros).

#### <span style="color:blue">A continuación se descarta la variable ID porque es primaria, Name tiene en su gran mayoría datos iguales, las variables Team y NOC son redundantes en su mayoría también así que solo se dejara Team y la variable Games porque es redundante, ya que las variables Year y Season tiene su contenido.

In [122]:
df = df.drop(columns=['ID', 'Name', 'NOC', 'Games']) 
df.head()

,Sex,Age,Height,Weight,Team,Year,Season,City,Sport,Event,Medal
0,M,24.0,180.0,80.0,China,1992,Summer,Barcelona,Basketball,Basketball Men's Basketball,NaN
1,M,23.0,170.0,60.0,China,2012,Summer,London,Judo,Judo Men's Extra-Lightweight,NaN
2,M,24.0,NaN,NaN,Denmark,1920,Summer,Antwerpen,Football,Football Men's Football,NaN
3,M,34.0,NaN,NaN,Denmark/Sweden,1900,Summer,Paris,Tug-Of-War,Tug-Of-War Men's Tug-Of-War,Gold
4,F,21.0,185.0,82.0,Netherlands,1988,Winter,Calgary,Speed Skating,Speed Skating Women's 500 metres,NaN


Gestione los datos duplicados y los datos nulos. En el caso de la variable **Medal**, impute a los datos nulos el valor **No medal**.

#### <span style="color:blue">Gestionando datos duplicados

In [123]:
print('Número de registros duplicados: ', df.duplicated().sum())

Número de registros duplicados:  11264


In [124]:
df = df.drop_duplicates()
print('Número de registros duplicados: ', df.duplicated().sum())

Número de registros duplicados:  0


#### <span style="color:blue">Gestionando datos nulos

In [125]:
df[['Medal', 'Weight', 'Height', 'Age']].isnull().sum()*100/df.shape[0]

Medal     85.255838
Weight    20.001770
Height    18.962717
Age        1.991903
dtype: float64

#### <span style="color:blue">Imputando No medal

In [126]:
df['Medal'].fillna(value='No medal').loc[37:45]

37      Bronze
38      Bronze
39    No medal
40      Bronze
41      Bronze
42        Gold
43    No medal
44        Gold
45    No medal
Name: Medal, dtype: object

#### <span style="color:blue">Analizando el porcentaje de los datos nulos que tienen las variables Wight, Height y Age, es bajo así que estos registros con datos nulos se pueden eliminar

In [127]:
df=df.dropna(subset=['Weight', 'Height', 'Age'])

## Análisis exploratorio de variables categóricas

Haga un análisis de las variables categóricas no descartadas, e identifique:

- Variables nominales.
- Variables ordinales.
- Variables con alta cardinalidad.
- Variables que no deberían incluirse en la matriz de características.
- Indiquen qué tipo de procesamiento debería recibir cada una de las variables categóricas, según su tipo y cardinalidad.

Reporte los resultados del análisis en una celda de texto.

In [128]:
df.describe(include='object').T

,count,unique,top,freq
Sex,205846,2,M,139212
Team,205846,660,United States,13662
Season,205846,2,Summer,166453
City,205846,42,London,13784
Sport,205846,56,Athletics,32354
Event,205846,590,Ice Hockey Men's Ice Hockey,3798
Medal,30132,3,Gold,10149


#### <span style="color:blue">-Variables nominales: Sex, Season, City y Sport

#### <span style="color:blue">-Variables ordinales: Medal

#### <span style="color:blue">-Variables con alta cardinalidad: Team, NOC y Event

#### <span style="color:blue">-Variables que no deberían incluirse en la matriz de características: Medal, City y Season

#### <span style="color:blue">-Tipo de procesamiento debería recibir cada una de las variables categóricas, según su tipo y cardinalidad:


## Análisis exploratorio de variables cuantitativas o numéricas

Haga un análisis de las variables cuantitativas no descartadas, e identifique:

- Variables que son aproximadamente normales.
- Variables con datos atípicos.
- Variables que no deberían incluirse en la matriz de características.
- Indique qué tipo de procesamiento debería recibir cada una de las variables numéricas, según su distribución y la presencia de valores atípicos.

Reporte los resultados del análisis en una celda de texto.

## Procesamiento de las características y entrenamiento del modelo

- Escoja un modelo de árbol de decisión que prediga la variable **Weight**, teniendo en cuenta que es numérica continua.
- Extraiga la matriz de características y el vector objetivo de su dataframe, a partir de los análisis previos.
- Parta los datos en subconjuntos de entrenamiento y de prueba.
- Procese las características del modelo usando `ColumnTransformer`.
- Entrene el modelo.
- Reporten los **scores** de entrenamiento y de prueba.
- Reporten el listado completo de características con las que fue entrenado el modelo después del procesamiento realizado.